### ISSUE TITLE
Add partial support for Objective-C. [phab:D1001]

### ISSUE THREAD

**Issue by kythe** [google/kythe#1335]
_Created by **salguarnieri** at 2016-08-08 21:18:32_: 

Add support for methods and interfaces. This does not include categories or
protocols. They require more work to fully represent in the graph. Instance
variables are also not supported yet. The todos in the code list many other
situations that are not yet fully implemented that are related to interfaces
or methods.

Todos are present for langauge features that require further investigation.

**Comment by kythe** [comment]
_**salguarnieri** wrote at 2016-08-08 23:02:32_:

A few updates before review

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-08-09 00:46:46_:

It might be better to use `expect_fail_verify = True` instead of commenting the test blocks out. Right now this test in particular fails to index (and I don't think that's the intention):

```
contravariant.runfiles/io_kythe/kythe/cxx/indexer/cxx/testdata/objc/generic_contravariant.m:9:10: error: use of undeclared identifier 'nil'
  return nil;
         ^
```

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-08-09 00:46:46_:

This I'm not sure of. If you *do* want to use a different `language` value, you'll need to be careful when generating cross-references to C code. To do that, you'll need to be able to determine for any given header file whether it's `objc` or `c++`.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-08-09 00:46:46_:

nit: should be a capital `Create`.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-08-09 00:46:46_:

nit: implementation

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-08-09 00:46:46_:

Is "super and protocols" missing a word? I'm unfamiliar with Objective-C so this might make perfect sense.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-08-09 00:46:46_:

nit: please add a Doxygen comment for this member function.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-08-09 00:46:46_:

nit (this and the below line):

```
if (isa<clang::CXXDestructorDecl>(Decl)) {
```

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-08-09 00:46:46_:

nit (this and below line):

```
if (auto *M = dyn_cast<clang::ObjCMethodDecl>(Decl)) {
```

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-08-09 00:46:46_:

I believe that you can get rid of this and the BuildNodeIdForDecl line below.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-08-09 00:46:46_:

Better to use `UNSUPPORTED_CLANG_TYPE(ObjCObjectPointer)` than leave a commented-out block.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-08-09 00:46:46_:

nit: this shouldn't have the "// Leaf" label, as it will recurse on `Impl`.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-08-09 00:46:46_:

Please use UNSUPPORTED_CLANG_TYPE here and below.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-08-09 00:46:46_:

Do you need separate `BodyDeclNode` and `DeclNode` IDs?

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-08-09 00:46:46_:

nit: commented-out code

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-08-09 00:46:46_:

nit: `if (const auto *SCTSI = ...)`

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-08-09 00:46:46_:

Do you need different `DeclNode` and `BodyDeclNode` IDs?

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-08-09 00:46:46_:

Do you need separate IDs?

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-08-09 00:46:46_:

This can go. I'm not sure which refactoring job left those dangling parent nodes around.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-08-09 00:46:46_:

Is there a distinction between InnerNode/OuterNode?

**Comment by kythe** [comment]
_**zarko** wrote at 2016-08-09 00:46:46_:

First couple of comments.

**Comment by kythe** [comment]
_**salguarnieri** wrote at 2016-08-10 22:43:23_:

Address comments.

**Comment by kythe** [comment]
_**salguarnieri** commented inline (context lost during migration) at 2016-08-10 22:47:46_:

Yeah, I ran into that during my thoughts here as well. It would simply some designs if we just called objective-c the same language name as c/c++.

**Comment by kythe** [comment]
_**salguarnieri** commented inline (context lost during migration) at 2016-08-10 22:47:46_:

It is to connect the current class to its super/parent class and the protocols the current class implements. Which is a mouthful.

ConnectToSuperClassAndImplementedProtocols?
ConnectToHierarchy?
ConnectToProtocolsAndSuperClass?
I'm open to other suggestions.

**Comment by kythe** [comment]
_**salguarnieri** commented inline (context lost during migration) at 2016-08-10 22:47:46_:

I believe I do not.

**Comment by kythe** [comment]
_**salguarnieri** commented inline (context lost during migration) at 2016-08-10 22:47:46_:

I think I am missing something because I don't see where the Impl is. TypeNodes.def doesn't show a type that is a subtype of ObjCInterface.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-08-11 00:43:43_:

I think that what's here is fine modulo using 'superclass' instead of 'super'.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-08-11 00:43:43_:

`ID = BuildNodeIdForDecl(Impl);`

(This was for TypeLoc::ObjCInterface, but it seems like the comment has drifted around.)

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-08-11 00:43:43_:

Sorry, `::ObjCInterface` looks fine; `::ObjCObjectPointer` does recurse on `OPTL.getPointeeLoc()`.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-08-11 00:43:43_:

nit: no space between `w` and `<` (please run clang-format over your C++ files and note that it does a bad job with `UNSUPPORTED_CLANG_TYPE`--but I'm not really that annoyed with it anymore, so feel free to accept its changes.)

Note that LLVM-style files should be formatted with -style=LLVM and Google-style files should be formatted with -style=Google.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-08-11 00:43:43_:

What span is associated with FooCall? I'd expect it to be `[box foo]`.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-08-11 00:43:43_:

If this callsite *always* calls FooDefn, it's better to record that; e.g., in C++, we have

```
//- D defines/binding FnD?
class C { public: void D(); };
//- D defines/binding FnDC?
//- D completes/uniquely FnD
void C::D() { }
void f(C& c) {
  //- "c.D()" ref/call FnDC
  c.D();
}
```

(we still pass this if `D` is virtual, and it's unclear to me whether this is the right thing logically; thoughts?)

**Comment by kythe** [comment]
_**salguarnieri** wrote at 2016-08-12 01:33:17_:

cleanup, better ranges for methoddecl and messageexpr

**Comment by kythe** [comment]
_**salguarnieri** commented inline (context lost during migration) at 2016-08-12 01:34:31_:

Well... I thought they would, but simple testing says they don't.

interface O
property int X
end

O *o; o.X = 4; // This calls [o setX: 4]

But it looks like that is directly modeled as a PropertyRefExpr and the message expr to setX is not marked as implicit.

**Comment by kythe** [comment]
_**salguarnieri** commented inline (context lost during migration) at 2016-08-12 01:34:31_:

So, probably the +(int) shouldn't be in there, but the notion that a method name has a single token is a little murky

+(int) foo:(int)a forInt:(int)i withChar:(char)c; //objc
vs int fooForIntWithChar(int a, int i, char c); // c

In C, we have a single token for the function name but in obj-c the selector (closest thing to name) would be foo:forInt:withChar:.

**Comment by kythe** [comment]
_**salguarnieri** commented inline (context lost during migration) at 2016-08-12 01:34:31_:

It is, made explicit in test.

**Comment by kythe** [comment]
_**salguarnieri** commented inline (context lost during migration) at 2016-08-12 01:34:31_:

Is that what c++ does now?

As a user I would want all of this information and I think it is all present in the graph.

We have the calls edge to the decl, we should have the overrides edges from the decl, and we should should record the type information to know which definition to go to. Actually, I don't have the type info in the graph yet (I think) but that should be possible.

So I'm not sure we need a direct call edge to the definition but I expect us to make use of path from callsite -> defn a lot.

**Comment by kythe** [comment]
_**salguarnieri** wrote at 2016-08-12 01:37:13_:

clang-format

**Comment by kythe** [comment]
_**zarko** accepted this revision at 2016-08-12 20:23:32_

**Comment by kythe** [comment]
_**salguarnieri** wrote at 2016-08-12 20:37:29_:

Rebased and resolved a merge conflict

**Comment by kythe** [comment]
_**salguarnieri** closed via commit 41da17cd8d8932d1bc975eac1e8c84fc2ee31d1a at 2016-08-12 20:49:39_
